doc/reverse_engineering/game_mechanics/formations-formulas.md

```diff
@@ -0,0 +1,248 @@
+
+        findNthLastBeforeInc(n) + 1 <= numOfUnits < findNthLastBeforeInc(n + 1)
+        
+with some constraints over `numOfUnits` and `n`.
+So, by doing:
+``` mathematica
+assumptions := 
+ n ∈ Integers && numOfUnits > 6 && numOfUnits ∈ Integers && n >= 0;
+
+Reduce[{findNthLastBeforeInc[n] + 1 <= numOfUnits < findNthLastBeforeInc[n + 1], assumptions}, n]
+```
+we find that (for `numOfUnits >= 16`), `n` must lie between 0 and the first root of a third-degree polynomial, namely:
+``` mathematica
+        Root[90 - 6 numOfUnits + 65 x + 12 x^2 + x^3 &, 1]
+```
+For `numOfUnits == 15`, `n` is `0`.
+Since we're looking for natural values, we just need the `ceil()` of this function of `numOfUnits`.
+The root can be shown explicitly using `ToRadicals[]`. Unfortunately, it turns out to be a very complex function, unsuitable for real-time computing. The plot of the values looks like that of a square-root-like function.
+The function
+
+``` 
+        a*x^(1/6) + b*x^(1/3) + c*x + d
+```
+
+with parameters
+
+``` mathematica
+{a -> -7.622169906683245, 
+b -> 3.6545500464975955, 
+c -> -0.002901063925120857, 
+d -> 4.129231118574412}
+```
+
+found using
+
+``` mathematica
+    FindFit[values, a x^(1/6) + b x^(1/3) + c*x + d, {a, b, c, d}, x]
+```
+
+may be a good choice.
+
+Finally, we rewrite our `findRowCount` as
+
+``` python
+import math
+
+def findRowCount(numOfUnits):
+    if numOfUnits <= 6:
+        return 1
+    if 7 <= numOfUnits and numOfUnits <= 15:
+        return 2
+
+    a = -7.622169906683245
+    b = 3.6545500464975955
+    c = -0.002901063925120857
+    d = 4.129231118574412
+    approx = a*pow(numOfUnits, 1.0/6) + b*pow(numOfUnits, 1.0/3) + c*numOfUnits + d
+    rowCount = math.ceil(approx) + 1
+    return (approx, rowCount)
+```
+
+# Improved algorithm
+
+If you look back at the experimental data, an educated guess may be that the points where the slopes change may be expressed out as `5*i - (i-1)`: for the n-th point we have 
+
+``` mathematica
+        Sum[(5 i - (i - 1)), {i, 1, n}] + 1
+``` 
+
+Mathematica solves this as 
+
+``` mathematica
+        1 + 3 n + 2 n^2
+``` 
+
+The first few values are `6, 15, 28, 45, 66`, which prove to be correct.
+
+Now we must find where our `numOfUnits` is greater or equal to the above
+
+``` mathematica
+        Reduce[1 + 3 n + 2 n^2 <= N, n, Integers]
+``` 
+
+And of course we take the upper bound of the interval found, namely 
+
+``` mathematica
+        -(3/4) + 1/4 Sqrt[1 + 8 N]
+``` 
+
+We need the next integer, so we `Ceil` the above,
+
+``` mathematica
+        Ceiling[-(3/4) + 1/4 Sqrt[1 + 8 N]]
+``` 
+
+The expression found is the `rowCount` of above, in a saner form.
+
+Again, the first row is
+
+``` python
+        rows[0] = math.ceil(numOfUnits / findRowCount(numOfUnits))
+```
+
+and the other rows are unchanged.
+
+The final Python code may be:
+
+``` python
+import math
+
+def findRowCount(numOfUnits):
+    return math.ceil((math.sqrt(1 + 8*numOfUnits) - 3) / 4)
+```
```

Comment:
This should be at the start of the formulas section. We don't need the unoptimized solutions.